io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs

```diff
@@ -0,0 +1,176 @@
+        Right resource -> case exitCase of
+          ExitCaseSuccess (Right b) -> runExceptT (release resource (ExitCaseSuccess b))
+          ExitCaseException e       -> runExceptT (release resource (ExitCaseException e))
+          _                         -> runExceptT (release resource ExitCaseAbort))
+      (either (return . Left) (runExceptT . use))
+    return $ do
+      -- The order in which we perform those two 'Either' effects determines
+      -- which error will win if they are both 'Left's. We want the error from
+      -- 'release' to win.
+      c <- ec
+      b <- eb
+      return (b, c)
+
+instance MonadMask m => MonadMask (ExceptT e m) where
+  mask f = ExceptT $ mask $ \u -> runExceptT $ f (q u)
+    where
+      q :: (m (Either e a) -> m (Either e a))
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+  uninterruptibleMask f = ExceptT $ uninterruptibleMask $ \u -> runExceptT $ f (q u)
+    where
+      q :: (m (Either e a) -> m (Either e a))
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+
+--
+-- WriterT instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (WriterT w m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (WriterT w m) where
+  catch (WriterT m) f = WriterT $ catch m (runWriterT . f)
+
+  generalBracket acquire release use = WriterT $ fmap f $
+      generalBracket
+        (runWriterT acquire)
+        (\(resource, w) e ->
+          case e of
+            ExitCaseSuccess (b, w') ->
+              g w' <$> runWriterT (release resource (ExitCaseSuccess b)) 
+            ExitCaseException err ->
+              g w  <$> runWriterT (release resource (ExitCaseException err))
+            ExitCaseAbort ->
+              g w  <$> runWriterT (release resource ExitCaseAbort))
+        (\(resource, w)   -> g w <$> runWriterT (use resource))
+    where f ((x,_),(y,w)) = ((x,y),w)
+          g w (a,w') = (a,w<>w')
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadMask m) => MonadMask (WriterT w m) where
+  mask f = WriterT $ mask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+  uninterruptibleMask f = WriterT $ uninterruptibleMask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+
+
+--
+-- RWST Instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (RWST r w s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (RWST r w s m) where
+  catch (RWST m) f = RWST $ \r s -> catch (m r s) (\e -> runRWST (f e) r s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = RWST $ \r s ->
```

Comment:
The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055102